pydantic_cli/shell_completion.py

```diff
@@ -0,0 +1,51 @@
+"""
+Loading of shtab to
+
+Because of how argparse's parser.exit works, It's not
+easy to reuse other actions. Specially the "eager" actions
+which can at any point in time call parser.exit() which
+doesn't work with how pydantic-cli is designed.
+"""
+import sys
+from argparse import Action, ArgumentParser
+from .argparse import TerminalEagerCommand
+
+SUPPORTED_SHELLS = ["zsh", "bash"]
+
+try:
+    import shtab  # type: ignore
+
+    HAS_AUTOCOMPLETE_SUPPORT = True
+except ImportError:
+    HAS_AUTOCOMPLETE_SUPPORT = False
+
+
+SUPPORTED_SHELLS = ["zsh", "bash"]
+
+
+class EmitShellCompletionAction(Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        if not HAS_AUTOCOMPLETE_SUPPORT:
+            raise ImportError("Unable to export to shell. shtab is not installed")
+
+        if values in SUPPORTED_SHELLS:
+            print(shtab.complete(parser, values))
+            sys.stderr.write("Completed writing shell output to stdout\n")
+            raise TerminalEagerCommand
+
+        raise ValueError(
+            f"Unsupported shell type ({values}. Supported shells {SUPPORTED_SHELLS} "
+        )
+
+
+def add_shell_completion_arg(
+    p: ArgumentParser, auto_complete_flag: str = "--emit-shell-completion"
+) -> ArgumentParser:
+    p.add_argument(
+        auto_complete_flag,
+        choices=list(SUPPORTED_SHELLS),
+        default=None,
+        help="Emit Shell Completion",
+        action=EmitShellCompletionAction,
+    )
+    return p
```

Comment:
this should all be simplifiable now with `shtab.add_argument_to()`

Reply:
Originally, I thought that I could reuse the Action in argparse. However, it's not really possible because "eager" actions such as `--help`, `--version` or `--emit-shell-completion` in `pydantic-cli` to raise `TerminalEagerCommand` as a control structure mechanism to get around argparse's odd "anyone (or action) can dump out of the program by calling parser.exit" odd design. This issue was raised in the original PEP, but was deemed not an issue for some reason. 

https://www.python.org/dev/peps/pep-0389/#discussion-sys-stderr-and-sys-exit

At any rate, this creates some friction points reusing the Action defined in `shtab`.